Debug-frontend-tests.md

```diff
@@ -0,0 +1,98 @@
+	    at _toBlob (core/templates/combined-tests.spec.js:92825:11)
+	    at _maybeConvertBody (core/templates/combined-tests.spec.js:92870:20)
+	    at TestRequest../node_modules/@angular/common/fesm5/http/testing.js.TestRequest.flush (core/templates/combined-tests.spec.js:92745:16)
+	    at UserContext.<anonymous> (core/templates/combined-tests.spec.js:61955:17)
+	    at ZoneDelegate../node_modules/zone.js/dist/zone.js.ZoneDelegate.invoke (core/templates/pages/about-page/about-page.module.js:237519:30)
+	    at ProxyZoneSpec../node_modules/zone.js/dist/proxy.js.ProxyZoneSpec.onInvoke (core/templates/combined-tests.spec.js:427781:43)
+	    at ZoneDelegate../node_modules/zone.js/dist/zone.js.ZoneDelegate.invoke (core/templates/pages/about-page/about-page.module.js:237518:36)
+	    at Zone../node_modules/zone.js/dist/zone.js.Zone.run (core/templates/pages/about-page/about-page.module.js:237276:47)
+	    at runInTestZone (core/templates/combined-tests.spec.js:427336:38)
+	    at UserContext.<anonymous> (core/templates/combined-tests.spec.js:427351:24)
+Chrome Headless 83.0.4103.116 (Mac OS 10.15.5): Executed 1 of 2277 (1 FAILED) (0 secs / 0.046 secs)
+```
+
+The `zone.js` stack elements relate to the [zone.js](https://github.com/angular/angular/tree/master/packages/zone.js) package that is handling our asynchronous operations. Notice that the line numbers are unreasonably large. This is because to run our tests, we load all our testing code and dependencies into `combined-tests.spec.js`. The line numbers refer to the resulting "mega-file," which doesn't actually exist on the file system.
+
+Below, we'll discuss some strategies you can use to overcome these difficulties. We suggest you follow a workflow like this:
+
+1. [Find the code for the failing test](#find-tests)
+2. [Run the test in isolation](#isolate-tests)
+3. Debug by [printing out debugging information](#verbose-mode), [investigating the code referenced by the stack trace](#find-stack-elements), and/or [running many times on CI](#run-tests-repeatedly-on-ci).
+
+## Find tests
+
+You can use test names to find test code files, but you can't just search for the whole name. You have to understand how test names are constructed. For example, here's the kind of test structure that produced the test name "Assets Backend API Service on dev mode should handle rejection when fetching a file fail" above:
+
+```js
+describe('Assets Backend API Service', () => {
+  describe('on dev mode', () => }
+    it('should handle rejection when fetching a file fails', () => {
+      ...
+    });
+  });
+});
+```
+
+The strings from the `describe` and `it` functions get concatenated with space (` `) delimiters to create the test name. The name of the `it` block usually starts with `should`, so we recommend searching the codebase for the portion of the test name from `should` to the end (e.g. "should handle rejection when fetching a file fails" in this case).
+
+## Isolate tests
```

Comment:
FYI, sometimes certain tests flake only when the previous tests run in a certain pattern. For example:

```
describe('', () => {
var test = 2;
beforeEach(() => {
...
});

it('test 1', () => {
test = 3;
});

it('test 2', () => {
expect(test).toBe(2);
})
});
```

in the above case if `test-1` runs before `test-2` only then will it fail. This might be a useful tip for developers

Reply:
done